harvesters/src/main/java/org/fao/geonet/kernel/harvest/harvester/simpleurl/RDFUtils.java

```diff
@@ -0,0 +1,228 @@
+            if (recordInfo != null) {
+                records.put(recordInfo.one(), recordInfo.two());
+            }
+        }
+
+        qe.close();
+        model.close();
+        return records;
+    }
+
+
+    private static Element checkForMissingRdfAbout(Element rdfModel) throws Exception {
+        List ns = rdfModel.getAdditionalNamespaces();
+        List<Element> nodeWithNoRdfAbout =
+            Xml.selectNodes(rdfModel,
+                ".//*[local-name() = 'Dataset' or local-name() = 'DataService'][not(@rdf:about)]", ns
+            );
+        nodeWithNoRdfAbout.forEach(n -> {
+            try {
+                Object httpIdentifier = Xml.selectSingle(n, "dct:identifier[matches(., '^http(s)://.*$')]/text()", ns);
+                if (httpIdentifier != null) {
+                    n.setAttribute("rdf:about", (String) httpIdentifier);
+                }
+            } catch (JDOMException e) {
+            }
+        });
+        return rdfModel;
+    }
+
+    private static Model checkAndCreateMissingCatalogRecords(Model model) throws IOException, URISyntaxException {
+        Query queryExtractNoRec = QueryFactory.create(getQueryString("extract-resources-no-records.rq"));
+        QueryExecution qe = QueryExecutionFactory.create(queryExtractNoRec, model);
+
+        Model newModel = model;
+        ResultSet resultIds = qe.execSelect();
+        while (resultIds.hasNext()) {
+            QuerySolution solution = resultIds.nextSolution();
+            newModel = createCatalogRecord(
+                newModel,
+                solution.get("resourceId").toString(),
+                solution.get("catalogId").toString()
+            );
+        }
+        return newModel;
+    }
+
+
+    private static InputStream getResourceAsStream(String resourcePath) {
+        return RDFUtils.class.getClassLoader().getResourceAsStream(resourcePath);
+    }
+
+    /**
+     * Convert a model back to RDF XML
+     * Only used for debugging
+     */
+    private static String toRdfString(Model model) {
+        StringWriter out = new StringWriter();
+        RDFDataMgr.write(out, model, Lang.RDFXML);
+        return out.toString();
+    }
+
+
+    private static String getQueryString(String queryFile) throws IOException, URISyntaxException {
+        return IOUtils.toString(getResourceAsStream(RESOURCE_FOLDER + "sparql/" + queryFile), Constants.CHARSET);
+    }
+
+    private static Model createCatalogRecord(Model model, String resourceId, String catalogId) throws IOException, URISyntaxException {
+        String recordUUID = resourceId;
+        Date now = new Date();
+        String localQuery = getQueryString("add-CatalogRecord.rq")
+            .replace("%recordID%", recordUUID)
+            .replace("%recordUUID%", recordUUID)
+            .replace("%resourceId%", resourceId)
+            // TODO: Should we set modified of catalog record to the date of publication?
```

Comment:
I agree we should find a fallback for the date to avoid updating the local records on every harvest.  
However, whether this is the publication date or the modified date of the dataset/service, it might not be provided or up to date.  
This is why we initially decided to always update local records when the date is not provided.   

To be discussed.   
Ideally, we want to move forward and expect harvested dcat feeds to provide the `dcat:CatalogRecord` directly. We might want to stop generating it when missing at some point in the future.